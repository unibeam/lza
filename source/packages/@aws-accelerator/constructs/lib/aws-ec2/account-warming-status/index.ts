/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * aws-ec2-pre-warm-account-status - lambda handler
 *
 * @param event
 * @returns
 */

import { SSMClient, PutParameterCommand } from '@aws-sdk/client-ssm';
import {
  EC2Client,
  DescribeInstancesCommand,
  DescribeVpcsCommand,
  DescribeSubnetsCommand,
  DeleteVpcCommand,
  DeleteSubnetCommand,
  TerminateInstancesCommand,
} from '@aws-sdk/client-ec2';
import { AdaptiveRetryStrategy } from '@aws-sdk/util-retry';
import { delay } from '@aws-accelerator/utils/lib/throttle';

const solutionId = process.env['SOLUTION_ID'] ?? '';
const retryStrategy = new AdaptiveRetryStrategy(() => Promise.resolve(5));
const ec2Client = new EC2Client({ customUserAgent: solutionId, retryStrategy });
const ssmClient = new SSMClient({ customUserAgent: solutionId, retryStrategy });

type InstanceDetails = {
  instanceId: string | undefined;
  minutesSinceLaunch: number | undefined;
};

//eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(event: any): Promise<
  | {
      IsComplete: boolean;
    }
  | undefined
> {
  console.log(event);
  const ssmPrefix: string = event.ResourceProperties['ssmPrefix'];
  const instanceDetails = await getInstanceDetails();

  // if no instance was found, it has been deleted return
  if (instanceDetails.instanceId === undefined) {
    return { IsComplete: true };
  }

  if (instanceDetails.minutesSinceLaunch! > 15) {
    console.log('Account warming time reached');
    await terminateInstance(instanceDetails.instanceId);
    await deleteVpc();
    await setSSMParameter(ssmPrefix, 'true');
    return {
      IsComplete: true,
    };
  }
  console.log('Waiting for warmup to complete');
  return { IsComplete: false };
}

async function getInstanceDetails(): Promise<InstanceDetails> {
  console.log('Getting Instance Id');
  const instanceDetails: InstanceDetails = { instanceId: undefined, minutesSinceLaunch: undefined };
  const ec2Instances = await ec2Client.send(
    new DescribeInstancesCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (ec2Instances.Reservations && ec2Instances.Reservations.length < 1) {
    return instanceDetails;
  }

  if (ec2Instances.Reservations![0].Instances?.length ?? 0 > 0) {
    instanceDetails.instanceId = ec2Instances.Reservations![0].Instances![0].InstanceId;
    const launchTime = ec2Instances.Reservations![0].Instances![0].LaunchTime;

    const currentTime = new Date();
    const msElapsed = currentTime.getTime() - launchTime!.getTime();
    instanceDetails.minutesSinceLaunch = msElapsed / 1000 / 60;
  }

  return instanceDetails;
}

async function deleteVpc() {
  console.log('Deleting VPC');
  const vpcId = await getVpcId();

  if (vpcId) {
    console.log('Deleting subnets');
    const subnets = await ec2Client.send(
      new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }),
    );
    for (const subnet of subnets.Subnets) {
      await ec2Client.send(new DeleteSubnetCommand({ SubnetId: subnet.SubnetId }));
    }
    console.log(`Deleting VPC with id: ${vpcId}`);
    await ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId }));
  }
}

async function getVpcId(): Promise<string | undefined> {
  console.log('Getting VPC Id');
  const vpcs = await ec2Client.send(
    new DescribeVpcsCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (vpcs.Vpcs?.length ?? 0 > 0) {
    return vpcs.Vpcs![0].VpcId;
  }
  return undefined;
}

async function setSSMParameter(ssmPrefix: string, parameterValue: string) {
  console.log('Updating SSM Parameter');
  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: `${ssmPrefix}/account/pre-warmed`,
        Value: parameterValue,
        Overwrite: true,
      }),
    );
  } catch (e) {
    console.log(e);
  }
}

async function terminateInstance(instanceId: string): Promise<void> {
  console.log(`Terminating EC2 instance: ${instanceId}`);
  await ec2Client.send(new TerminateInstancesCommand({ InstanceIds: [instanceId] }));
  await waitForTermination(instanceId);
}

async function waitForTermination(instanceId: string): Promise<void> {
  console.log(`Waiting for termination of instanceId: ${instanceId}`);
  let ec2Terminated = false;
  while (!ec2Terminated) {
    const statusResponse = await ec2Client.send(new DescribeInstancesCommand({ InstanceIds: [instanceId] }));
    if (
      statusResponse.Reservations[0].Instances[0].State?.Name !== 'terminated' ||
      statusResponse.Reservations[0].Instances[0].State?.Code !== 48
    ) {
      delay(15000);
      continue;
    }
    ec2Terminated = true;
  }
}
