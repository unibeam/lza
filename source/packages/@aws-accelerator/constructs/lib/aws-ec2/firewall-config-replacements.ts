/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import path from 'path';
import { LzaCustomResource } from '../lza-custom-resource';

interface FirewallConfigReplacementProps {
  /**
   * Custom resource CloudWatch Log encryption key, when undefined default AWS managed key will be used
   */
  readonly cloudWatchLogKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource CloudWatch Log retention
   */
  readonly cloudWatchLogRetentionInDays: number;
  /**
   * Custom resource environment encryption key, when undefined default AWS managed key will be used
   */
  readonly environmentEncryptionKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource properties
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly properties: { [key: string]: any }[];
  /**
   * Custom resource role
   */
  readonly role: cdk.aws_iam.IRole;
}

export class FirewallConfigReplacements extends cdk.Resource {
  constructor(scope: Construct, id: string, props: FirewallConfigReplacementProps) {
    super(scope, id);

    new LzaCustomResource(this, 'Resource', {
      resource: {
        name: 'Resource',
        parentId: id,
        properties: props.properties,
      },
      lambda: {
        assetPath: path.join(__dirname, 'firewall-config-replacements/dist'),
        description: 'Firewall configuration replacement custom resource',
        environmentEncryptionKmsKey: props.environmentEncryptionKey,
        cloudWatchLogKmsKey: props.cloudWatchLogKey,
        cloudWatchLogRetentionInDays: props.cloudWatchLogRetentionInDays,
        role: props.role,
        timeOut: cdk.Duration.seconds(120),
      },
    });
  }
}
