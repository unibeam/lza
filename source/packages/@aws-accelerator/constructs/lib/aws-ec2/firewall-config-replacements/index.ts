/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { EC2Client } from '@aws-sdk/client-ec2';
import {
  CopyObjectCommand,
  CopyObjectCommandOutput,
  GetObjectCommand,
  ListObjectsV2Command,
  PutObjectCommand,
  PutObjectCommandOutput,
  S3Client,
} from '@aws-sdk/client-s3';
import { FirewallReplacements, VpnConnectionProps, initReplacements } from './replacements';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';

export interface IStaticReplacements {
  /**
   * The key name for the replacement
   */
  readonly key: string;
  /**
   * The value for the replacement
   */
  readonly value: string;
}

export interface FirewallReplacementOptions {
  /**
   * The AWS Partition
   */
  readonly partition: string;
  /**
   * The name of the S3 asset bucket
   */
  readonly assetBucketName: string;
  /**
   * The name of the S3 config bucket
   */
  readonly configBucketName: string;
  /**
   * The VPC ID where the firewall resides
   */
  readonly vpcId: string;
  /**
   * The key name of the configuration file
   */
  readonly configFileKey?: string;
  /**
   * The directory of the configuration files
   */
  readonly configDir?: string;
  /**
   * The hostname of the firewall instance
   */
  readonly firewallName?: string;
  /**
   * The instance ID of the firewall instance
   */
  readonly instanceId?: string;
  /**
   * The key name of the license file
   */
  readonly licenseFileKey?: string;
  /**
   * The name of the role to assume to retrieve cross-account values
   */
  readonly roleName?: string;
  /**
   * Static replacements
   */
  readonly staticReplacements?: IStaticReplacements[];
  /**
   * VPN connection details to look up
   */
  readonly vpnConnectionProps?: VpnConnectionProps[];
  /**
   * Management Account ID used to read Secrets Manager secrets for replacements
   */
  readonly managementAccountId?: string;
}

export async function handler(event: CloudFormationCustomResourceEvent): Promise<{ Status: string } | undefined> {
  //
  // Set custom resource options
  const options = setOptions(event.ResourceProperties);
  //
  // Set up clients
  const ec2Client = new EC2Client({ customUserAgent: process.env['SOLUTION_ID'] });
  const s3Client = new S3Client({ customUserAgent: process.env['SOLUTION_ID'] });
  //
  // Begin handler logic
  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      //
      // Copy license file
      await copyLicenseFile(s3Client, options);
      //
      // Process config file replacements
      const replacements =
        options.configFileKey || options.configDir
          ? await initReplacements(ec2Client, event.ServiceToken, options)
          : undefined;
      await processConfigFileReplacements(s3Client, options, replacements);

      return {
        Status: 'SUCCESS',
      };

    case 'Delete':
      //
      // Do nothing
      return {
        Status: 'SUCCESS',
      };
  }
}

/**
 * Set firewall replacement options based on event
 * @param resourceProperties { [key: string]: any }
 * @returns FirewallReplacementOptions
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function setOptions(resourceProperties: { [key: string]: any }): FirewallReplacementOptions {
  return {
    partition: (resourceProperties['ServiceToken'] as string).split(':')[1],
    assetBucketName: resourceProperties['assetBucketName'] as string,
    configBucketName: resourceProperties['configBucketName'] as string,
    vpcId: resourceProperties['vpcId'] as string,
    configFileKey: (resourceProperties['configFile'] as string) ?? undefined,
    configDir: (resourceProperties['configDir'] as string) ?? undefined,
    firewallName: (resourceProperties['firewallName'] as string) ?? undefined,
    instanceId: (resourceProperties['instanceId'] as string) ?? undefined,
    licenseFileKey: (resourceProperties['licenseFile'] as string) ?? undefined,
    roleName: (resourceProperties['roleName'] as string) ?? undefined,
    staticReplacements: (resourceProperties['staticReplacements'] as IStaticReplacements[]) ?? undefined,
    vpnConnectionProps: (resourceProperties['vpnConnections'] as VpnConnectionProps[]) ?? undefined,
    managementAccountId: (resourceProperties['managementAccountId'] as string) ?? undefined,
  };
}

/**
 * Copy license file from asset bucket to config bucket
 * @param s3Client S3Client
 * @param options FirewallReplacementOptions
 * @returns Promise<CopyObjectCommandOutput | undefined>
 */
async function copyLicenseFile(
  s3Client: S3Client,
  options: FirewallReplacementOptions,
): Promise<CopyObjectCommandOutput | undefined> {
  if (!options.licenseFileKey) {
    return;
  }

  console.log(
    `Copying license file ${options.licenseFileKey} from bucket s3://${options.assetBucketName} to s3://${options.configBucketName}...`,
  );
  try {
    const response = await throttlingBackOff(() =>
      s3Client.send(
        new CopyObjectCommand({
          CopySource: `${options.assetBucketName}/${options.licenseFileKey}`,
          Bucket: options.configBucketName,
          Key: options.licenseFileKey,
        }),
      ),
    );

    return response;
  } catch (e) {
    throw new Error(`${e}`);
  }
}

/**
 *
 * @param s3Client S3Client
 * @param options FirewallReplacementOptions
 * @param replacements string | undefined
 * @returns Promise<CopyObjectCommandOutput | undefined>
 */
async function processConfigFileReplacements(
  s3Client: S3Client,
  options: FirewallReplacementOptions,
  replacements?: FirewallReplacements,
): Promise<PutObjectCommandOutput | undefined> {
  //
  // Validate input
  if ((!options.configFileKey || !options.configDir) && !replacements) {
    return;
  } else if ((options.configFileKey || options.configDir) && !replacements) {
    throw new Error(`Configuration file/directory S3 key provided but replacements are undefined`);
  } else if ((options.configFileKey || options.configDir) && replacements) {
    const configFiles = options.configFileKey ? [options.configFileKey] : [];
    if (options.configDir) {
      const objs = await s3Client.send(
        new ListObjectsV2Command({ Bucket: options.assetBucketName, Prefix: options.configDir }),
      );
      for (const obj of objs.Contents ?? []) {
        // Empty folders are ignored here
        if (obj.Key!.endsWith('/')) continue;
        configFiles.push(obj.Key!);
      }
    }
    for (const configFileKey of configFiles) {
      //
      // Get raw config asset
      const rawFile = await getRawConfigFile(s3Client, options.assetBucketName, configFileKey);
      //
      // Process replacements
      const transformedFile = await transformConfigFile(replacements, rawFile);
      //
      // Put transformed config file
      await putTransformedConfigFile(s3Client, options.configBucketName, transformedFile, configFileKey);
    }
  }
  return;
}

/**
 * Get raw configuration file from asset bucket
 * @param s3Client S3Client
 * @param assetBucketName string
 * @param configFileKey string
 * @returns Promise<string | undefined>
 */
async function getRawConfigFile(
  s3Client: S3Client,
  assetBucketName: string,
  configFileKey: string,
): Promise<string | undefined> {
  console.log(`Retrieving raw configuration file ${configFileKey} from bucket s3://${assetBucketName}...`);
  try {
    const response = await throttlingBackOff(() =>
      s3Client.send(new GetObjectCommand({ Bucket: assetBucketName, Key: configFileKey })),
    );

    return await response.Body?.transformToString();
  } catch (e) {
    throw new Error(`${e}`);
  }
}

/**
 * Transform configuration file with replacements
 * @param replacements FirewallReplacements
 * @param configFile string | undefined
 * @returns string
 */
async function transformConfigFile(replacements: FirewallReplacements, configFile?: string): Promise<string> {
  if (!configFile) {
    throw new Error(`Encountered an error retrieving configuration file from S3`);
  }
  //
  // Process config file replacements
  const lookupRegex = /\${ACCEL_LOOKUP::(EC2|CUSTOM|SECRETS_MANAGER)(:.[^}]+){1,3}}/gi;
  const variables = [...new Set(configFile.match(lookupRegex) ?? [])];
  const replacedVariables = await replacements.processReplacements(variables);
  //
  // Transform variables to regex
  const regexVariables = transformVariablesToRegex(variables);

  for (const [index, value] of regexVariables.entries()) {
    configFile = configFile.replace(new RegExp(value, 'g'), replacedVariables[index]);
  }
  return configFile;
}

/**
 * Transform variables to a regex pattern
 * @param variables string[]
 * @returns string[]
 */
function transformVariablesToRegex(variables: string[]): string[] {
  const regexVariables: string[] = [];

  for (const variable of variables) {
    let regexVariable = variable;
    regexVariable = regexVariable.replace(/\$/, '\\$');
    regexVariable = regexVariable.replace(/\{/, '\\{');
    regexVariable = regexVariable.replace(/\}/, '\\}');
    regexVariables.push(regexVariable);
  }
  return regexVariables;
}

/**
 * Put transformed config file to config bucket
 * @param s3Client
 * @param configBucketName
 * @param transformedFile
 * @param configFileKey
 * @returns Promise<PutObjectCommandOutput | undefined>
 */
async function putTransformedConfigFile(
  s3Client: S3Client,
  configBucketName: string,
  transformedFile: string,
  configFileKey: string,
): Promise<PutObjectCommandOutput | undefined> {
  console.log(`Putting transformed configuration file ${configFileKey} to bucket s3://${configBucketName}...`);
  try {
    return await throttlingBackOff(() =>
      s3Client.send(new PutObjectCommand({ Bucket: configBucketName, Body: transformedFile, Key: configFileKey })),
    );
  } catch (e) {
    throw new Error(`${e}`);
  }
}
