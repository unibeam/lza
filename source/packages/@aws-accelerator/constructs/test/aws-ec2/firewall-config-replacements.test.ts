/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { FirewallConfigReplacements } from '../../lib/aws-ec2/firewall-config-replacements';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(FirewallConfigReplacements): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

// Create mock keys and role for custom resource
const cloudWatchKey = new cdk.aws_kms.Key(stack, 'CloudWatchKey');
const lambdaKey = new cdk.aws_kms.Key(stack, 'LambdaKey');
const role = new cdk.aws_iam.Role(stack, 'Role', {
  assumedBy: new cdk.aws_iam.ServicePrincipal('lambda.amazonaws.com'),
});

// Create resource
new FirewallConfigReplacements(stack, 'ConfigReplacements', {
  cloudWatchLogKey: cloudWatchKey,
  cloudWatchLogRetentionInDays: 3653,
  environmentEncryptionKey: lambdaKey,
  properties: [{ test: 'test' }],
  role,
});

describe('FirewallConfigReplacements', () => {
  snapShotTest(testNamePrefix, stack);
});
