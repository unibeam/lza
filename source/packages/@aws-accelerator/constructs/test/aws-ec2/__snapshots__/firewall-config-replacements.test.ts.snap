// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirewallConfigReplacements Construct(FirewallConfigReplacements):  Snapshot Test 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Resources": {
    "CloudWatchKey9B181885": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "ConfigReplacementsFunction3B8BED87": {
      "DependsOn": [
        "Role1ABCC5F0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Firewall configuration replacement custom resource",
        "Handler": "index.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "LambdaKey984A39D9",
            "Arn",
          ],
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "Role1ABCC5F0",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConfigReplacementsFunctionResourceLogGroupA52DC796": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CloudWatchKey9B181885",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "ConfigReplacementsFunction3B8BED87",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "ConfigReplacementsResourceResource58F31BA3": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "ConfigReplacementsFunctionResourceLogGroupA52DC796",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "ConfigReplacementsframeworkonEvent7670032B",
            "Arn",
          ],
        },
        "test": "test",
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "ConfigReplacementsframeworkonEvent7670032B": {
      "DependsOn": [
        "ConfigReplacementsframeworkonEventServiceRoleDefaultPolicy8E284567",
        "ConfigReplacementsframeworkonEventServiceRole6796BD33",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/ConfigReplacements/Resource/Resource)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "ConfigReplacementsFunction3B8BED87",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "ConfigReplacementsframeworkonEventServiceRole6796BD33",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConfigReplacementsframeworkonEventServiceRole6796BD33": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "AWS Custom resource provider framework-role created by cdk.",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ConfigReplacementsframeworkonEventServiceRoleDefaultPolicy8E284567": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Allows only specific policy.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ConfigReplacementsFunction3B8BED87",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ConfigReplacementsFunction3B8BED87",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ConfigReplacementsframeworkonEventServiceRoleDefaultPolicy8E284567",
        "Roles": [
          {
            "Ref": "ConfigReplacementsframeworkonEventServiceRole6796BD33",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaKey984A39D9": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "Role1ABCC5F0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
